# IA - PROJETO 2
ALUNO: Flávio Henrique Simões Brito
## RELATÓRIO
### QUESTÃO 1

In [5]:
# Importar bibliotecas necessárias
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Carregar a base de dados
url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/housing.csv"
column_names = ["CRIM", "ZN", "INDUS", "CHAS", "NOX", "RM", "AGE", "DIS", "RAD", "TAX", "PTRATIO", "B", "LSTAT", "MEDV"]
data = pd.read_csv(url, header=None, names=column_names)
data = data.dropna(axis=0) # remover valores nulos

# Escolher uma variável preditora (Exemplo: 'RM') e a variável alvo ('MEDV')
X = data[['RM']].values
y = data[['MEDV']].values

# Dividir os dados em treinamento e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 1. Modelo de Regressão Linear
linear_model = LinearRegression()
linear_model.fit(X_train, y_train)
y_pred_linear = linear_model.predict(X_test)

# Avaliar o modelo de regressão linear
print("Regressão Linear:")
print(f"Mean Absolute Error: {mean_absolute_error(y_test, y_pred_linear):.2f}")
print(f"Mean Squared Error: {mean_squared_error(y_test, y_pred_linear):.2f}")

# 2. Modelo de Rede Neural
nn_model = Sequential([
 Dense(10, activation='relu', input_shape=(1,)),
 Dense(1, activation='linear')
])
nn_model.compile(optimizer='adam', loss='mse', metrics=['mae'])
# Treinar o modelo de rede neural
nn_model.fit(X_train, y_train, epochs=50, batch_size=10, verbose=1)
# Avaliar o modelo de rede neural
y_pred_nn = nn_model.predict(X_test)
print("Rede Neural:")
print(f"Mean Absolute Error: {mean_absolute_error(y_test, y_pred_nn):.2f}")
print(f"Mean Squared Error: {mean_squared_error(y_test, y_pred_nn):.2f}")

Regressão Linear:
Mean Absolute Error: 4.48
Mean Squared Error: 46.14
Epoch 1/50


/home/flavio/.local/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


41/41 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - loss: 722.5514 - mae: 25.0530   
Epoch 2/50
41/41 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 628.3442 - mae: 23.3316
Epoch 3/50
41/41 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 602.0506 - mae: 22.7923
Epoch 4/50
41/41 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 574.4435 - mae: 22.1264 
Epoch 5/50
41/41 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 511.4047 - mae: 20.5865
Epoch 6/50
41/41 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 487.1001 - mae: 20.0378
Epoch 7/50
41/41 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 443.8762 - mae: 18.9131
Epoch 8/50
41/41 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 377.8200 - mae: 17.1764
Epoch 9/50
41/41 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 325.0967 - mae: 16.0261
Epoch 10/50
41/41 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 308.3550 - mae: 15.2900 
Epoch 11/50
41/41 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 252.2767 - mae: 13.4346
Epoch 12/50
41/41 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 230.8318 - mae: 12.4269
Epoch 13/50
41/41 ━

#### RESULTADOS
1. Regressão Linear:
    - mean absolute error: 4.48
    - mean squared error: 46.14
2. Rede Neural: 
    - mean absolute error: 5.30
    - mean squared error: 56.51
### QUESTÃO 2

In [ ]:
# codigo questao 2